extension/src/time_weighted_average.rs

```diff
@@ -0,0 +1,422 @@
+                (Some(state), None) => Some(state),
+                (Some(state), Some(next)) =>  Some(state.combine(&next.to_TimeWeightSummary()).unwrap().into())
+            }
+        })
+    }
+}
+
+#[pg_extern(schema = "timescale_analytics_experimental")]
+fn time_weight_ordered_final(
+    state: Option<Internal<TimeWeightSummary>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<time_weight_summary<'static>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            let state = match state {
+                None => return None,
+                Some(state) => state,
+            };
+
+            flatten!(
+                time_weight_summary {
+                    method: &state.method,
+                    first: &state.first,
+                    last: &state.last,
+                    w_sum: &state.w_sum,
+                }
+            ).into()
+        })
+    }
+}
+
+extension_sql!(r#"
+CREATE OR REPLACE FUNCTION time_weight_summary_in(cstring) RETURNS time_weight_summary IMMUTABLE STRICT PARALLEL SAFE LANGUAGE C AS 'MODULE_PATHNAME', 'time_weight_summary_in_wrapper';
+CREATE OR REPLACE FUNCTION time_weight_summary_out(time_weight_summary) RETURNS CString IMMUTABLE STRICT PARALLEL SAFE LANGUAGE C AS 'MODULE_PATHNAME', 'time_weight_summary_out_wrapper';
+
+CREATE TYPE time_weight_summary (
+    INTERNALLENGTH = variable,
+    INPUT = time_weight_summary_in,
+    OUTPUT = time_weight_summary_out,
+    STORAGE = extended
+);
+
+CREATE AGGREGATE time_weight_ordered(method text, ts timestamptz, value DOUBLE PRECISION)
+(
+    sfunc = time_weight_ordered_trans,
+    stype = internal,
+    finalfunc = time_weight_ordered_final,
+    combinefunc = time_weight_ordered_combine,
+    serialfunc = time_weight_summary_serialize,
+    deserialfunc = time_weight_summary_deserialize,
+    parallel = restricted
+);
+
+CREATE AGGREGATE time_weight_ordered(tws time_weight_summary)
+(
+    sfunc = time_weight_summary_ordered_trans,
+    stype = internal,
+    finalfunc = time_weight_ordered_final,
+    combinefunc = time_weight_ordered_combine,
+    serialfunc = time_weight_summary_serialize,
+    deserialfunc = time_weight_summary_deserialize,
+    parallel = restricted
+);
+"#);
+
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct TimeWeightTransState {
+    #[serde(skip)]
+    buffer: Vec<TSPoint>,
+    method: Option<TimeWeightMethod>,
+    combine_buffer: Vec<TimeWeightSummary>,
+}
+
+impl TimeWeightTransState {
+    fn push(&mut self, value: TSPoint, method: TimeWeightMethod) {
+        match self.method {
+            None => {self.method = Some(method)},
+            Some(m)=> {if m != method {panic!("Mismatched methods")}}
+        }
+        self.buffer.push(value);
+    }
+
+    fn sort_and_calc(&mut self) {
+        if self.buffer.is_empty() {
+            return
+        }
+        self.buffer.sort_unstable_by_key(|p| p.ts);
```

Comment:
It seems like we're trying to tolerate data being entered out of order here.  However it looks like this will be in an invalid state if anything in the buffer has a timestamp earlier than the last timestamp present when we last called this function.  Is that a problem?  If we do find ourselves in that state we won't actually blow up until the combine_sorted_iter call in `combine_sort_and_calc` later on, which isn't great.

Reply:
Yeah, we will error in that case, but I'm not sure it's worth it to check, as there's any number of ways we could be in an invalid state here, no two summaries can have overlapping state, so checking that would require scanning the whole thing. It's also relatively unlikely that we will end up with two calls to that function anyway.

Reply:
(But I'm going to add comments to clarify some of this, because it is quite confusing)